pyo3-bytes/src/bytes.rs

```diff
@@ -62,6 +63,66 @@ impl PyBytes {
     pub fn as_slice(&self) -> &[u8] {
         self.as_ref()
     }
+
+    /// Slice the underlying buffer using a Python slice object
+    ///
+    /// This should behave the same as Python's byte slicing:
+    ///     - `ValueError` if step is zero
+    ///     - Negative indices a-ok
+    ///     - If start/stop are out of bounds, they are clipped to the bounds of the buffer
+    ///     - If start > stop, the slice is empty
+    ///
+    /// This is NOT exposed to Python under the `#[pymethods]` impl
+    fn slice(&self, slice: &Bound<'_, PySlice>) -> PyResult<PyBytes> {
+        let len_isize = self.0.len() as isize;
+        let psi = slice.indices(len_isize)?;
+        let (start, stop, step) = (psi.start, psi.stop, psi.step);
+        if step == 0 {
+            return Err(PyValueError::new_err("step is zero"));
+        }
+
+        // I think this is right!?
+        let new_cap_usize = if (step > 0 && stop > start) || (step < 0 && stop < start) {
+            (((stop - start).abs() + step.abs() - 1) / step.abs()) as usize
+        } else {
+            0
+        };
+
+        if new_cap_usize == 0 {
+            return Ok(PyBytes(Bytes::new()));
+        }
+
+        // if start < 0  and stop > len and step == 1 just copy?
+        if step == 1 && start < 0 && stop >= len_isize {
+            let out = self.0.slice(..);
+            let py_bytes = PyBytes(out);
+            return Ok(py_bytes);
+        }
+
+        if step == 1 && start >= 0 && stop <= len_isize && start < stop {
+            let out = self.0.slice(start as usize..stop as usize);
+            let py_bytes = PyBytes(out);
+            return Ok(py_bytes);
+        }
```

Comment:
I think it would be clearer to combine these cases of `step` into a single if/else case block

```rs
if step == 0 {

} else if step == 1 {
 .. above code
} else if step > 0 {
let mut new_buf = BytesMut::with_capacity(new_cap_usize);
...
} else if step < 0 {
...
} else {
	unreachable!("Value of step: {}", step)
}
```

It's fine to always compute the new capacity before checking if `step == 0`, because we assume that's an error if it is passed in.

Reply:
@kylebarron fixed this... I dont think we actually have to check for step==0, as the `.indices` call on pyslice throws that error for us.